### ISSUE TITLE
[Google Cloud Shell] Screen Reader causing full page crashes when terminal is updated frequently.

### ISSUE THREAD

**Issue by pattch** [xtermjs/xterm.js#3374]
The current screen reader implementation waits for render frames in order to debounce refreshes in the accessibility tree. This can cause layout thrashing in pages that have enough elements with enough styles applied, etc. when terminals are updated frequently without newlines being entered. Even worse, when layout thrashing becomes a bottleneck for processing content to be updated in the accessibility tree, the internal character buffer used by xterm's screen reader will grow indefinitely, eventually causing out of memory errors.

This issue is not seen when debouncing in time domain independent of render frames (e.g. debouncing to update the screen reader every one second as opposed to every render frame), as well as when putting a limit on the maximum size of the character buffer. Since the screen reader doesn't need to be updated as frequently as animated output, this seems like a fair tradeoff.

Additionally, this issue is not seen when the screen reader is disabled altogether, using e.g. `screenReaderMode: false`.

Example:
In Google Cloud Shell, running long-running gsutil commands would cause page performance to degrade, due to the fast spinner being used to indicate progress over a long period of time in xterm Terminals. Eventually, this would result in an out of memory error, crashing the entire page.

## Details
- Browser and browser version: Chrome, Firefox verified; assume this issue effects all browsers.
- OS version: Verified across macOS, linux, and Windows.
- xterm.js version: Synced at commit 5cb048094e23da22ccfe46666b87d1ad427dce6d

![Screen Shot 2021-06-09 at 11 06 28 AM](https://user-images.githubusercontent.com/6233815/123311334-ea5fb800-d4db-11eb-8ad6-918e2214c65e.png)
*Example profile of xterm terminal causing layout thrashing, not seen when the screen reader is disabled.*

**Comment by jerch** [comment]
> In Google Cloud Shell, running long-running gsutil commands would cause page performance to degrade, due to the fast spinner being used to indicate progress over a long period of time in xterm Terminals. Eventually, this would result in an out of memory error, crashing the entire page.

Well, this is bad. And this only happens with screen reader enabled, or does the memory always keep growing? Just to rule out other reasons like missing data flow control...

My first guess here is in fact - missing flow control. Normally with flow control the screen reader updates should just align with data transport sequentially (blocking each other) and stalling further data delivery, thus a page crash should not be possible. Without flow control a page can get overwhelmed by data, which would lead to tons of scheduled work on the event loop eventually crashing things.

Your image above looks more than toxic and Ive never seen anything like that. Can you also share the runtime table of JS functions? Furthermore the JS heap usage is very high with 82 MB - are you sure nothing else on that page causes all this? For me xterm.js in the demo never gets beyond 15 MB allocated stuff there.

**Comment by pattch** [comment]
> Well, this is bad. And this only happens with screen reader enabled, or does the memory always keep growing? Just to rule out other reasons like missing data flow control...

With the screen reader enabled, if the site is very simple this doesn't actually occur. Only if the page is complicated enough to have the layout thrashing does the memory issue present itself. On pages that are complex enough (Google Cloud Shell for example) with many elements and styles applied, the character buffer inside of xterm's screen reader will continue to grow. Concisely, this only happens when the screen reader is on, on complex pages, with very fast output to xterm terminals that doesn't include newline characters; without all of those conditions, the screen reader can normally keep up and clear out the internal buffer. With the screen reader disabled (as it is now on Google Cloud Shell in production), there is no memory issue.

One way to replicate this issue (on sufficiently complicated layouts) is by using the following script, assuming the xterm terminal backed by some real VM / similar:

```
spin='-\|/'

i=0
while true
do
        i=$(( (i+1) %4 ))
        printf "\r${spin:$i:1}"
done
```

> My first guess here is in fact - missing flow control. Normally with flow control the screen reader updates should just align with data transport sequentially (blocking each other) and stalling further data delivery, thus a page crash should not be possible. Without flow control a page can get overwhelmed by data, which would lead to tons of scheduled work on the event loop eventually crashing things.

We've patched this on our end by simply debouncing using a fixed amount of time (e.g. one second) instead of render frames, which alone is enough to prevent the screen reader's character buffer from growing continuously. A fixed cap on the character buffer would just be another added layer of safety.

> Your image above looks more than toxic and I've never seen anything like that. Can you also share the runtime table of JS functions? Furthermore the JS heap usage is very high with 82 MB - are you sure nothing else on that page causes all this? For me xterm.js in the demo never gets beyond 15 MB allocated stuff there.

I agree this is really bad, and the heap is much smaller (and never grows) when the screen reader is disabled. In fact, with the screen reader disabled, the above script can run just fine for hours (updating very frequently, etc.) without causing any performance degradation, memory increase, etc.

Like I said, this is hard to reproduce without a sufficiently complex page (I tried building a dummy terminal to repro, but in isolation this issue isn't hit), but when it does hit it can cause the entire page to crash (Aw Snap in Google, a simple Out of Memory error in Firefox).

**Comment by jerch** [comment]
@pattch Thx for the script, can repro it in the demo with screenreader enabled. This looks more than wrong:

![grafik](https://user-images.githubusercontent.com/6193135/123327576-dd25e600-d53a-11eb-9939-f35a4641b74d.png)

Around 3s I switched screenreader on, from there on something gets stacked up, and the layouting takes longer and longer with each iteration.

> ... with very fast output to xterm terminals that **doesn't include newline characters**; without all of those conditions, the screen reader can normally keep up and clear out the internal buffer.

Maybe that's our issue behind it? Looks like something clear buffers on newline chars only, but keeps growing without. Would at least explain the ever growing layouting time (if all that ends up in the DOM).

@Tyriar Any thoughts on this?

**Comment by jerch** [comment]
@pattch Found the culprit - without a newline in between, things are added up in the DOM here: https://github.com/xtermjs/xterm.js/blob/094bcbd81d1f080d204331088aa99f4d7173d56c/src/browser/AccessibilityManager.ts#L297

This grows as long as no NL is sent eventually leading to a crash. A quickfix might be some ringbuffer logic here:
```typescript
const MAX_ANNOUNCE = 4096;

  private _announceCharacters(): void {
    if (this._charsToAnnounce.length === 0) {
      return;
    }
    if (this._liveRegion.textContent  && this._liveRegion.textContent.length > MAX_ANNOUNCE) {
      this._liveRegion.textContent = this._liveRegion.textContent.slice(this._charsToAnnounce.length) \
        + this._charsToAnnounce;
    } else {
      this._liveRegion.textContent += this._charsToAnnounce;
    }
    this._charsToAnnounce = '';
  }
```
But I did not write the screenreader part, thus dont know if it would defeat the purpose of the `_liveRegion`. Another idea would just stop propagating further data until it got cleared by an NL.

**Comment by Tyriar** [comment]
Good find!

If it hits some large maximum like that then it could just be cleared, instead of bothering with a ring buffer approach as that would probably cause issues with announcement anyway.

**Comment by pattch** [comment]
Another issue is that the accessibility tree is updating very frequently, which causes the layout thrashing (which degrades performance enough to cause the memory issue in the accessibility manager). I think it is also important to throttle the updates to the screen reader instead of debouncing on animation frames as is done using the RenderDebouncer today:

https://github.com/xtermjs/xterm.js/blob/094bcbd81d1f080d204331088aa99f4d7173d56c/src/browser/RenderDebouncer.ts#L29-L62

**Comment by Tyriar** [comment]
Good idea, updating so frequently doesn't help screen reader users anyway so we could do it via `RenderDebouncer` so that we stick to an animation frame but throttle to a second or so? Created https://github.com/xtermjs/xterm.js/issues/3376

**Comment by jerch** [comment]
@Tyriar  Are there any blocking constraints on those accessibility interfaces? Like "all has to be seen by the accessibility application"? I have no clue about the rules here, just wonder how a webapp should behave, if the consuming device takes much longer, like a braille display has a max baud rate and such. What if I input `ls -l /` and get 100 pages of text flowing, how would such a braille display handle that? Shouldn't the baud rate limit/block the flow in the first place?

**Comment by Tyriar** [comment]
We already limit what's shown when too many _lines_ are output at once, this would just extend that to long lines as well https://github.com/xtermjs/xterm.js/blob/094bcbd81d1f080d204331088aa99f4d7173d56c/src/browser/AccessibilityManager.ts#L223

So we truncate the data read when there's too much, but when the user hits that message they can always go back and read line by line.

**Comment by jerch** [comment]
> We already limit what's shown when too many lines are output at once, this would just extend that to long lines as well

But doesn't that mean, that our accessibility support is basically broken for blocking output devices, as in case of braille displays? :confused:

**Comment by Tyriar** [comment]
I don't have experience with braille displays but I don't see why they would differ in this regard to screen readers

**Comment by jerch** [comment]
Well I have no experience at all with accessibility/aria stuff, thus no informed opinion about it. Still withholding information seems wrong to me (also for screen readers). To my understanding highly dynamic content is a problem for any screen reader/accessibility technology, and discouraged by the guidelines (did only a quick reading). Ofc a terminal is all about a stream of text, pretty much the opposite of 98% of other computer use cases, where you can navigate within document boundaries. A screen reader perfectly fits the document case, but not our dynamic "text window". To be more in line with accessibility guidelines, I think we would need a slow/page scroll mode, so a screen reader or braille display can keep up.

Since most of that is speculative for me, I wont put too much thinking into it. So I am fine with either way.

**Comment by pattch** [comment]
> @Tyriar Are there any blocking constraints on those accessibility interfaces? Like "all has to be seen by the accessibility application"? I have no clue about the rules here, just wonder how a webapp should behave, if the consuming device takes much longer, like a braille display has a max baud rate and such. What if I input ls -l / and get 100 pages of text flowing, how would such a braille display handle that? Shouldn't the baud rate limit/block the flow in the first place?

> This is somewhat linked to the debouncing idea - with a lower update rate we would withhold/skip information, thus a blind user would not get all the data. How to go about that?

> Well I have no experience at all with accessibility/aria stuff, thus no informed opinion about it. Still withholding information seems wrong to me (also for screen readers). To my understanding highly dynamic content is a problem for any screen reader/accessibility technology, and discouraged by the guidelines (did only a quick reading). Ofc a terminal is all about a stream of text, pretty much the opposite of 98% of other computer use cases, where you can navigate within document boundaries. A screen reader perfectly fits the document case, but not our dynamic "text window". To be more in line with accessibility guidelines, I think we would need a slow/page scroll mode, so a screen reader or braille display can keep up.

When there is a ton of content being displayed to the screen with newlines in between, the Screen Reader implementation today has no problems keeping up. In fact, on newline characters, the entire buffer gets announced and from that point on if it's too much content for screen readers, each one has their own way of handling large amounts of content as far as I'm aware.

The bug as reported only presents when a large amount of content is rapidly written to the Terminal without newline characters (e.g. a progress bar or spinner displaying progress). I can't see any issue with truncating / throttling this kind of content for screen readers, since it's abnormal, even for Terminals.

**Comment by jerch** [comment]
@pattch Sorry for extending your bug to this - what I am trying to address here seems to be a fundamental flaw in how we present data to screen readers:
I dont really doubt that screen readers cannot deal with the amount of data xterm.js may produce (we are typically in the 5 - 35 MB/s range). Screen reader should be fine grabbing that much data per time, at least as native apps.
But - there is not really a good way to present that much data to impaired users, if the data keeps flowing, even for screen readers. Thus any accessibility software would have to throttle data presented, so peeps can follow. Screen readers prolly have interfaces for that, but we dont offer any hooks to slow down or temporarily block data flow. Thats all I am concerned about here with my remarks.
I cannot really help to solve that, as I have no idea about the inner works of screen readers beside some conceptual reading about it. And that tells me - we create a problem for those users, if we swallow genuine content.

I totally agree that a graphical spinner would create tons of nonsense data for a screen reader. But reading the guidelines closely I'd say they are not accessibility friendly - they misuse graphical terminal caps of forth and back updates to simulate progress. I know that tons of tools use them for their fanciness, but that only shows that ppl dont think much about accessibility in general and the implications for other output means.

**Comment by pattch** [comment]
Just wanted to update here as well: I have an open PR that fixes this issue without running into the issues mentioned before:

> But - there is not really a good way to present that much data to impaired users, if the data keeps flowing, even for screen readers. Thus any accessibility software would have to throttle data presented, so peeps can follow. Screen readers prolly have interfaces or conventions for that, but we dont offer any hooks to slow down or temporarily block data flow. Thats all I am concerned about here with my remarks.

The PR I've mentioned (#3399 ) doesn't trim any content, only throttling how frequently the accessibility tree is updated in time as opposed to updating every render cycle.

In addition, just as a side note, the existing Accessibility reader already trims content announced to screen readers, but it does so in terms of rows as opposed to the absolute character count. If there's 20 or more rows, the existing Accessibility reader will announce "Too much content to announce. Navigate to rows to read content" or something along those lines.

If there's any issue with the PR I've opened for this, please let me know
